src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -3340,6 +3332,73 @@ private RubyNode addNewlineIfNeeded(ParseNode jrubyNode, RubyNode node) {
         return node;
     }
 
+    private interface ValueFromNode {
```

Comment:
So this seems a mechanism to avoid the extra local variable when the `receiverNode` is a SelfParseNode, so that visitCallNode() can find out if the receiver is just `self`?
This took me a while to understand.
Could you add a comment explaining that?

Is it actually correct for all SideEffectFree?
It doesn't seem to be for the translation of e.g. ComplexParseNode, which generates a new instance on every execution, as then it will invoke the reader and writer methods on different instances.
So I would limit it and simplify to `SelfParseNode`, which inside a method is guaranteed to always return the same object/instance.

I wonder if an extra `boolean receiverIsSelf` on `CallParseNode` wouldn't be simpler though. OTOH it seems nice to avoid the extra local variable.

Reply:
> So this seems a mechanism to avoid the extra local variable when the receiverNode is a SelfParseNode, so that visitCallNode() can find out if the receiver is just self?

Yes.

I'll limit it.

Reply:
Fixed.